### PR TITLE
Mfs26

### PR DESCRIPTION
**Pull request by andrewbird**
Here's the new int2f/111e/5f06 function and user converted. One thing though in `FindRedirectionByDevice()` I have buffer size specified with a define, but unless I add size parameters and convert its callers, I don't know how to deal with it?

### CODE REVIEW

File: src/dosext/mfs/mfs.c [hunk 1]
```diff
@@ -2433,16 +2433,17 @@ path_to_dos(char *path)
     *s = '\\';
 }
 
-static int GetRedirection(struct vm86_regs *state)
+static int GetRedirection(struct vm86_regs *state, ushort subfunc)
```
Comment (stsp) [inline src/dosext/mfs/mfs.c:2436]: Or, rather, pass len?

**Comment by stsp** [comment]
> I have buffer size specified with a define, but unless I add size parameters and convert its callers,
> I don't know how to deal with it?

How about just asserting that the passed value
does not exceed such define?
Also 256 looks too small. Why not, say, 1K?

File: src/dosext/mfs/mfs.c (hunk 1 above)
Comment (andrewbird) [reply src/dosext/mfs/mfs.c]: Done, looks tidier.

**Comment by andrewbird** [comment]
I increased the buffer to 1024, though I'm not sure how those share paths will be created.
Not sure I understand about the proposed switch to assert in `FindRedirectionByDevice()` as it is passed a char pointer for resourceStr, only its caller knows the length of the buffer.

**Comment by stsp** [comment]
> though I'm not sure how those share paths will be created.

What share paths?

> Not sure I understand about the proposed switch to assert

`assert(resourceSize <= MAX_RESOURCE_SIZE);`
(which is 1024)
Just to make sure no one allocates
from DOS too much.

**Comment by andrewbird** [comment]
> What share paths?

Well to see any advantage in lredir output of 5f06 over 5f05 you'd need to create a share mapping of \\LINUX\VeryLongPathOfOver128Chars. From what I can see 5f03 can only handle 128 chars share resource input, see http://www.ctyme.com/intr/rb-3043.htm , so as we use that function to create shares both internally and via lredir, I don't see how any longer share can be present, do you?

>assert(resourceSize <= MAX_RESOURCE_SIZE);
(which is 1024)

MAX_RESOURCE_PATH_LENGTH is 128, there's no such define as MAX_RESOURCE_SIZE

**Comment by stsp** [comment]
> so as we use that function to create shares

Not sure why you call that "shares".
If you have seen that word somewhere in a
samba config, then I am pretty sure it meant
something else. For example, directory exported
via smb protocol, so that multiple clients can
share it over the network.

> I don't see how any longer share can be present, do you?

No idea what "longer share" is, but yes, I see
what you mean.
However, consider this:
1. 5f05/5f06 allows larger buffers, so there must be the use
2. 5f03 mentions that the name should be small,
but it tells nothing about the buffer, because the
buffer is supplied _and filled_ by client.

So if client filled a large buffer with a larger name,
then the server has the right to either return an
error, or process the name correctly. It has all
the info to do so, so he can just ignore that note
about 128 bytes. It doesn't influence anything,
its mostly just a note.

**Comment by stsp** [comment]
For example if it is said that
"client supplies the buffer of 128bytes
max and server fills it" then indeed you
can't go around the limit.
But when client supplies the filled buffer,
server can process it and nothing bad
will happen.

**Comment by andrewbird** [comment]
> Not sure why you call that "shares".

Because Microsoft invented the UNC which is of the form `\\[server]\[sharename]\` that's what we are supposed to be returning in the resource field of 5f02/5/6

> 5f05/5f06 allows larger buffers, so there must be the use

Does it? rbil mentions that 05 has BP set to the Netbios session ID, but no mention of bigger buffers. 06 is undocumented, at least as far as I can find.

> No idea what "longer share" is

Dosemu's redirector presents a path derived from the unix path as if it were the sharename, longer path = longer sharename.

> For example if it is said that
"client supplies the buffer of 128bytes
max and server fills it" then indeed you
can't go around the limit.
But when client supplies the filled buffer,
server can process it and nothing bad
will happen.

Agreed, but just be sure that DOS itself doesn't copy the buffer in the processing from int21/5f03 to int2f/111e/5f03. I'm pretty sure at least one DOS I tested a while back did this.

**Comment by stsp** [comment]
> Because Microsoft invented the UNC which is of the form \\[server]\[sharename]\

Yes but you said `we use that function to create shares`.
If you want to apply this to the UNC convention,
then you can say that we create shares by `$_lredir_paths`
as after that, the "share" became shared, i.e. accessible
by client. And the aforementioned function is used to
mount these shares, rather than to create.
But I wouldn't apply that terminology here anyway, as
we do not have network or SMB here.

> Does it?

```
               5F02h  get redirection list entry
                        BX = redirection list index
                        DS:SI -> 16-byte local device name buffer
                        ES:DI -> 128-byte network name buffer
...
                5F05h  get redirection list extended entry
                        BX = redirection list index
                        DS:SI -> buffer for ASCIZ source device name
                        ES:DI -> buffer for destination ASCIZ network path
```

Explicit sizes were replaced with ASCIZ, so if
you ask me - yes it does, but RBIL did not
reverse-engineered the hard-coded maximum
lengths. They are not written, and are not passed,
either. Where are they? So clearly it does, but the
description is incomplete.

> at least one DOS I tested a while back did this.

Which?
Sad news.

**Comment by stsp** [comment]
> MAX_RESOURCE_PATH_LENGTH is 128, there's no such define as MAX_RESOURCE_SIZE

But you can add it to check that the
caller doesn't request too much.

**Comment by stsp** [comment]
> at least one DOS I tested a while back did this.

Well, at worst the large names will return an
error on that DOS. Its not a regression.

**Comment by andrewbird** [comment]
> Explicit sizes were replaced with ASCIZ, so if
you ask me - yes it does, but RBIL did not
reverse-engineered the hard-coded maximum
lengths.

I see what you mean, but as you say it only makes sense if the client passes the buffer size. I had a quick look at that MSClient 3.0 5f05 dosdebug session log I pasted in https://github.com/dosemu2/dosemu2/issues/1259#issuecomment-678659346 , CX is set to 0x0800(2048) there, so that may be passing the buffer size. I was hoping to confirm this with the Breadbox Ensemble client log here #710, but unfortunately it used 5f02. Other potential 5f05 clients are PC-NFS and Pathworks, but I don't think I ever got them setup on Dosemu due to lack of the correct type of network card.
I'm wondering if we really need to guess at what 5f06 is, and instead merely pass buffer length in CX for 5f05? We could check for ludicrous values of CX and adjust to 128, otherwise transfer up to the size it asks for.

>>    at least one DOS I tested a while back did this.

>Which?

Sorry I can't remember, I only noticed because I was trying to check the es:di passed in against biosseg:offset and it was different from the int21/5f03 callers es:di. obtained via the stack.

**Comment by stsp** [comment]
> I'm wondering if we really need to guess at what 5f06 is

That's just a safest bet.

> and instead merely pass buffer length in CX for 5f05? We could check for ludicrous values of
> CX and adjust to 128, otherwise transfer up to the size it asks for.

You've already seen 2048 - is it an
artificial value or valid?
So you can do `if (CX > 2048) assume_128();`,
but in a range of 128...2048 you crash.
Another possibility is to check the client's
signature, but note that checking the stored
signature, as I suggested previously, is not
enough in this case, because the client that
list redirections, may be a different one
(like ensamble is). So checking the signature
is perfectly fine, but its currently in CX.

So... what about moving the signature
into DH first?

**Comment by andrewbird** [comment]
So I updated to pass the signature in DX, and so only use CX for buffer size if the client is Dosemu. Interestingly I noticed another awkwardness in the RBIL description of int2f/111e/5f05:

5F05h  get redirection list extended entry
BX = redirection list index
DS:SI -> buffer for ASCIZ source device name **There is no size defined for this either, so what should a redirector transfer?**
ES:DI -> buffer for destination ASCIZ network path

I also wonder about why the terms source and destination are used, as both are for output?

**Comment by stsp** [comment]
No, I didn't mean like that. :)
Maybe we could move signature to DH
globally, and then use 0x5f05 instead
of 0x5f06.

> There is no size defined for this either, so what should a redirector transfer?

I think this one doesn't increase though.
16 should be enough.

> I also wonder about why the terms source and destination are used, as both are for output?

Clearly they meant local/remote instead.

**Comment by stsp** [comment]
I moved the signature check to DX, enjoy.
It was quite a PITA to compactify everything
and update all callers, but seems to work.

**Comment by andrewbird** [comment]
I rebased on top of your new stuff. Not sure if you want to use 5f05 or 5f06, but I have to go out now for the day.

**Comment by stsp** [comment]
You have said:
`I'm wondering if we really need to guess at what 5f06 is, and instead merely pass buffer length in CX for 5f05?`
I am trying to make it so you can
do just that. With signature being
in DX, its much easier to do it your
way!

**Comment by stsp** [comment]
> I rebased on top of your new stuff.

Still conflicts.

**Comment by stsp** [comment]
Applied with edits.

**Pull request closed without merging**